standards/application/messaging-api.md

```diff
@@ -0,0 +1,944 @@
+the appropriate events from the `message:*` event family (as described in the `Event Source API` section) MUST be invoked after the message is added or updated in the underlying storage.
+
+#### Acknowledgements
+
+Message Acknowledgement mechanism is a background operation that MUST:
+- Perform store‑based reliability checks as specified in [P2P‑RELIABILITY](./p2p‑reliability.md).  
+- Halt when the node’s health status transitions to `unhealthy` (as defined by the `Health API`).  
+- Resume when the node’s health status transitions to either `minimally healthy` or `healthy`.  
+- Set the `stored` or `received` field on `MessageRecord` once the message is acknowledged according to store‑based reliability.
+
+This operation MUST be initiated when any of the following conditions is met:
+- The `storeConfirmation` or `receiveConfirmation` property is `true` and `confirmContentTopics` is provided.  
+- The `Subscribe API` is invoked for one or more content topics.  
+- The `Send API` has successfully sent one or more messages.
+
+For the recurring [STORE](../standards/core/store.md) queries it is RECOMMENDED to prioritize nodes in this order:
+1. `staticStoreNode` specified in the initial configuration.
+2. `preferredServiceNodes`, if provided.
+3. Nodes discovered via network discovery.
+
+#### Programmatic API
+
+```typescript
+type MessageRecord = {
+  sending?: boolean;
+  sent?: boolean;
+  stored: boolean;
+  received: boolean;
+  requestId?: RequestID;
+  message: IWakuMessage;
+  error?: unknown;
+};
+
+interface MessageStorage {
+  getByRequestId(id: RequestID): MessageRecord | undefined;
+  getByHash(hash: string, skip?: number, take?: number): MessageRecord | undefined;
+  getByContentTopic(contentTopic: string, skip?: number, take?: number): MessageRecord[];
+  getBySubscriptionId(id: SubscriptionID, skip?: number, take?: number): MessageRecord[];
+}
+```
+
+#### Properties of `MessageRecord`
+
+##### `sending`
+Defaults to `false`.
+Indicates whether a message, originating from the underlying node,
+is currently in the process of being transmitted via the `Send API`.
+This property is set to `true` while the message is actively being sent or during retry attempts.
+Once the send operation completes (either successfully or if retry attempts are halted),
+the property MUST transition back to `false`.
+This property is not applicable to messages that are received by the node.
+
+##### `sent`
+Defaults to `false`.
+This field is applicable only to messages sent by the underlying node.
+Indicates whether a message, sent by the underlying node via the `Send API`,
+was successfully transmitted.
+Once a message is sent successfully, the `sent` property becomes `true` and remains unchanged.
+If the transmission fails, the property transitions to `false` and the `error` field is populated with implementation-specific details.
+This update occurs only after the `sending` property has transitioned back to `false`.
```

Comment:
From what heuristics?

I think some clarification is needed.

I'd say:

Indicates whether a message was sent over-the-wire with a best effort heuristic depending on the outbound protocols used. In the case of Light Push, an OK response from the remote service node will set this property to `true`.
For relay, a seemingly healthy pool of peers will sent this to `true` as soon as the message has been written on the wire.